genesyscloud/architect_grammar/genesyscloud_architect_grammar_proxy.go

```diff
@@ -0,0 +1,283 @@
+			log.Printf("Dealing with grammar id : %s", *grammar.Id)
+			allGrammars = append(allGrammars, grammar)
+		}
+	}
+
+	return &allGrammars, nil
+}
+
+// getArchitectGrammarByIdFn is an implementation of the function to get a Genesys Cloud Architect Grammar by Id
+func getArchitectGrammarByIdFn(ctx context.Context, p *architectGrammarProxy, grammarId string) (grammar *platformclientv2.Grammar, statusCode int, err error) {
+	grammar, resp, err := p.architectApi.GetArchitectGrammar(grammarId, true)
+	if err != nil {
+		return nil, resp.StatusCode, fmt.Errorf("Failed to retrieve grammar by id %s: %s", grammarId, err)
+	}
+	return grammar, resp.StatusCode, nil
+}
+
+// getArchitectGrammarIdByNameFn is an implementation of the function to get a Genesys Cloud Architect Grammar by name
+func getArchitectGrammarIdByNameFn(ctx context.Context, p *architectGrammarProxy, name string) (grammarId string, retryable bool, err error) {
+	const pageNum = 1
+	const pageSize = 100
+	grammars, _, err := p.architectApi.GetArchitectGrammars(pageNum, pageSize, "", "", []string{}, name, "", "", true)
+	if err != nil {
+		return "", false, fmt.Errorf("Error searching architect grammar %s: %s", name, err)
+	}
+
+	if grammars.Entities == nil || len(*grammars.Entities) == 0 {
+		return "", true, fmt.Errorf("No architect grammars found with name %s", name)
+	}
+
+	var grammar platformclientv2.Grammar
+	for _, grammarSdk := range *grammars.Entities {
+		if *grammarSdk.Name == name {
+			log.Printf("Retrieved the grammar id %s by name %s", *grammarSdk.Id, name)
+			grammar = grammarSdk
+			return *grammar.Id, false, nil
+		}
+	}
+
+	return "", false, fmt.Errorf("Unable to find grammar with name %s", name)
+}
+
+// updateArchitectGrammarFn is an implementation of the function to update a Genesys Cloud Architect Grammar
+func updateArchitectGrammarFn(ctx context.Context, p *architectGrammarProxy, grammarId string, grammar *platformclientv2.Grammar) (*platformclientv2.Grammar, error) {
+	grammar, _, err := p.architectApi.PatchArchitectGrammar(grammarId, *grammar)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to update grammar %s: %s", grammarId, err)
+	}
+
+	return grammar, nil
+}
+
+// updateArchitectGrammarLanguageFn is an implementation function for updating a Genesys Cloud Architect Grammarlanguage
+func updateArchitectGrammarLanguageFn(ctx context.Context, p *architectGrammarProxy, grammarId string, languageCode string, language *platformclientv2.Grammarlanguage) (*platformclientv2.Grammarlanguage, error) {
+	// Need to check if the language exists, if not, create it
+	_, _, err := p.architectApi.GetArchitectGrammarLanguage(grammarId, languageCode)
+	if err != nil {
+		log.Printf("Laguage %s not found, creating language", languageCode)
+		language, err := p.createArchitectGrammarLanguage(ctx, grammarId, language)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create grammar language: %s", err)
+		}
+
+		return language, nil
+	}
```

Comment:
Is there a specific scenario when this would be needed? Seems like terraform would always be calling the create function for you before an update

Reply:
Ya without this I was getting an error when I try to add a new language to a grammar after it is created. Since grammar languages use their own endpoints the language has to exist to update it

Reply:
I wonder if it would be better to have grammar languages as another resource? I see there's a Create, Read, Update and Delete function in the SDK anyway.

Reply:
I did consider this at the start but thought it would be unnecessary. It could be needed though because it's not actually possible to delete a language once it's added since it is it's own object. User prompts behaves the same